design/pipeline-caching.md

```diff
@@ -0,0 +1,154 @@
+
+| Experiment # | Execution Time | Approach                                                 |
+| ------------ | -------------- | -------------------------------------------------------------- |
+| 1            | 47 seconds     | ```yarn``` with clean cache                                    |
+| 2            | 19 seconds     | ```yarn``` with dirty cache                                    |
+| 3            | 47 seconds     | dedup file transfer with ```tar``` extraction                  |
+| 4            | 22 seconds     | dedup file transfer with direct file placement                 |
+| 5            | 44 seconds     | dedup file transfer with direct file placement into yarn cache |
+| 6            | 42 seconds     | dedup file transfer with ```tar``` & gzip compression          |
+
+The timings above are relative and performed in a local test environment. They are shown here to establish the impact that selecting different caching strategies can have on performance.
+
+To establish a baseline of performance experiment #1 just executes the ```yarn``` command in a workspace with no ```node_modules``` folder and a clean Yarn cache - then in experiment #2 we similuate the local development workstation scenario by removing the ```node_modules``` folder by maintain the Yarn cache directory.
+
+For experiment #3 we used a dedup file transfer mechanism built into Azure Artifacts (used behind the scenes by Universal Packages and Pipeline Artifacts). As part of the experiment we packed the node_modules folder into a tar file (without compression) which resulted in a 150MB file. We were able to transfer that file in about 19 seconds and then unpacking took about another 28 seconds - resulting in the same performance as a Yarn install on a clean cache (no improvement).
+
+In experiment #4 we eliminated the tar file and allowed the dedup file transfer to directly place files on disk. This process took 22 seconds end to end which roughly halves the Yarn installation time.
+
+In experiment #5 we instead cached the Yarn package cache directory instead of the local ```node_modules``` folder. The download of the packages into the Yarn cache directory took 22 seconds similar to experiment #4, however Yarn installation added 22 seconds for linking which meant it was roughly the same as a clean install. My comparison, running Yarn install with experiment #4 is close to a no-op for Yarn.
+
+In order to explore the archive file scenario a little bit more we did experiment #6 where in addition to creating a ```.tar``` archive, we also added compression to the file (```.tgz```). Due to file size reductions the transfer time was reduced to about 7 seconds, but decompression and placing files using ```tar``` took 35 seconds.
```

Comment:
As decompression + placing files is the activity that took the most time, it might be worth experimenting with more modern compression algorithms that decompress faster, while still retaining decent compression ratios. [Zstandard](https://facebook.github.io/zstd/) is a good candidate.